### PR TITLE
Adding support for automatically skipping cancelation if the amount l…

### PR DESCRIPTION
**Pull request by garrettmoon**
…eft to be downloaded is less than the average time to first byte.

Putting this up in favor of #357

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/323f5b5dccb36e8a336581b1d047d9d3/raw/556deff66ee4aae9abd400992916a18cd031990c/log.txt)

File: Source/Classes/PINURLSessionManager.m [hunk 1]
```diff
@@ -187,6 +193,60 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
...
+    NSDate *requestStart = [NSDate distantFuture];
+    NSDate *firstByte = [NSDate distantPast];
+    
+    BOOL valid = YES;
+    
+    for (NSURLSessionTaskTransactionMetrics *metric in metrics.transactionMetrics) {
+        if (metric.requestStartDate == nil || metric.responseStartDate == nil) {
+            //Only evaluate requests which completed their first byte.
+            valid = NO;
+            break;
```
Comment (jparise) [inline Source/Classes/PINURLSessionManager.m]: Seems like you could drop the `valid` flag and just `return` here.

File: Source/Classes/PINURLSessionManager.m [hunk 2]
```diff
@@ -187,6 +193,60 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
...
+        [_timeToFirstByteCache setObject:existingTimeToFirstByte forKey:host];
+    [self unlock];
+}
+
+- (NSTimeInterval)weightedTimeToFirstByteForHost:(NSString *)host
+{
+    NSTimeInterval timeToFirstByte;
+    [self lock];
+        timeToFirstByte = [[_timeToFirstByteCache objectForKey:host] doubleValue];
+        if (timeToFirstByte <= 0 + DBL_EPSILON) {
```
Comment (jparise) [inline Source/Classes/PINURLSessionManager.m]: You can drop the lock earlier (here).

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/c70ff00c93327308e7c303806e7d5149/raw/3732b147dd55b3e094ade56b67c8754646741bba/log.txt)

File: Source/Classes/PINURLSessionManager.m [hunk 3]
```diff
@@ -187,6 +193,60 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
...
+    }
+    
+    if (valid) {
+        [self storeTimeToFirstByte:[firstByte timeIntervalSinceDate:requestStart] forHost:task.originalRequest.URL.host];
+    }
+}
+
+- (void)storeTimeToFirstByte:(NSTimeInterval)timeToFirstByte forHost:(NSString *)host
+{
+    [self lock];
```
Comment (jparise) [inline Source/Classes/PINURLSessionManager.m]: I wonder if you can get away without locking. `NSCache` supports concurrent access, and the logic below is sufficiently fuzzy that missing/overwriting a measurement probably won't be super meaningful.

**Pull request merged**